WordPress/gutenberg
# Create Block: Support Custom Prompts in External Templates

---

Closes https://github.com/WordPress/gutenberg/issues/47524

---



## What?
This PR enhances the `@wordpress/create-block` package by allowing external project templates to define custom prompts, enabling developers to specify additional parameters during block or plugin creation.

## Why?
Currently, the `@create-block` package limits users to a predefined set of prompts, restricting customization based on specific project needs. Allowing custom prompts provides greater flexibility and aligns the tool with diverse development requirements.

## How?
The implementation involves modifying the `getDefaultValues` and `getPrompts` functions within the `templates.js` file to recognize and incorporate custom parameters defined in external templates. This ensures that during the scaffolding process, any additional prompts specified are presented to the user and their inputs are integrated into the generated project files.

## Testing Instructions
1. Prepare an external project template that includes custom prompt configurations.
2. Use the modified `@create-block` package to generate a new block or plugin with the custom template.
3. Verify that the custom prompts are displayed during the creation process.
4. Confirm that the responses to the custom prompts are accurately reflected in the scaffolded project files.

### Testing Instructions for Keyboard
1. Initiate the block or plugin creation process using the terminal.
2. Navigate through the prompts using keyboard inputs.
3. Ensure all prompts, including custom ones, are accessible and can be answered solely via the keyboard.

## Screenshots or screencast

This quick demo contains a couple of examples of how I'd envision this working:

https://github.com/user-attachments/assets/763d3296-f86f-41a0-bf11-1109bcb4dbde

<details><summary>Heres the code for: my-template/index.js</summary>
<p>

```js
/**
 * Dependencies
 */
const { join } = require("path");

module.exports = {
    defaultValues: {
        title: "My Template",
        description: "My Template Description",
        icon: "admin-post",
        // Default values for custom prompts
        analyticsId: 'UA-11111-2',
        colorScheme: 'default',
        enableAdvancedFeatures: false,
        customCssPrefix: 'default-prefix-',
    },
    customPrompts: {
        analyticsId: {
            type: 'input',
            name: 'analyticsId',
            message: 'Enter your Google Analytics ID (optional, format: UA-XXXXX-Y):',
            validate(input) {
                if (!input) return true; // Optional
                if (!/^UA-\d+-\d+$/.test(input)) {
                    return 'Invalid Google Analytics ID format. Should be like UA-XXXXX-Y';
                }
                return true;
            }
        },
        colorScheme: {
            type: 'list',
            name: 'colorScheme',
            message: 'Choose a color scheme for your block:',
            choices: [
                { name: 'Default Theme Colors', value: 'default' },
                { name: 'Custom Palette', value: 'custom' },
                { name: 'Monochrome', value: 'mono' }
            ],
            default: 'default'
        },
        enableAdvancedFeatures: {
            type: 'confirm',
            name: 'enableAdvancedFeatures',
            message: 'Would you like to enable advanced block features?',
            default: false
        },
        customCssPrefix: {
            type: 'input',
            name: 'customCssPrefix',
            message: 'Enter a custom CSS prefix for your block (optional):',
            validate(input) {
                if (!input) return true; // Optional
                if (!/^[a-z][a-z0-9-]*$/.test(input)) {
                    return 'CSS prefix must start with a letter and contain only lowercase letters, numbers, and hyphens';
                }
                return true;
            },
            filter(input) {
                return input.toLowerCase();
            }
        }
    },
    // Make sure template files are found
    pluginTemplatesPath: join(__dirname, 'files', 'plugin'),
    blockTemplatesPath: join(__dirname, 'files', 'block'),
};
```

</p>
</details> 


This enhancement addresses the need for customizable prompts in the `@create-block` package, as discussed in issue [#47524](https://github.com/WordPress/gutenberg/issues/47524).
